src/utils/daemon.rs

```diff
@@ -4,13 +4,49 @@
 
 use crate::{
     utils::{queue_builder, report_error},
-    Context, RustwideBuilder,
+    BuildQueue, Config, Context, Index, RustwideBuilder,
 };
 use anyhow::{anyhow, Context as _, Error};
 use log::{debug, info};
+use std::sync::Arc;
 use std::thread;
 use std::time::{Duration, Instant};
 
+pub fn watch_registry(
+    build_queue: Arc<BuildQueue>,
+    config: Arc<Config>,
+    index: Arc<Index>,
+) -> Result<(), Error> {
+    let mut last_gc = Instant::now();
+
+    // On startup we fetch the last seen index reference from
+    // the database and set it in the local index repository.
+    if let Some(oid) = build_queue.last_seen_reference()? {
+        index.diff()?.set_last_seen_reference(oid)?;
+    }
+
+    loop {
+        if build_queue.is_locked()? {
+            debug!("Queue is locked, skipping checking new crates");
+        } else {
+            debug!("Checking new crates");
+            match build_queue
+                .get_new_crates(&index)
+                .context("Failed to get new crates")
```

Comment:
Can you add a comment that there should only be one registry watcher running at a time? It confused me for a bit why there wasn't any locking.

In practice I think it should be ok - the worst that happens is we'll try and insert the same crates twice, and the second insert will be ignored by the database.

Reply:
I added a note